samples/Mvc.DynamicCorsPolicyProvider/DynamicCorsPolicyProvider.cs

```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Cors.Infrastructure;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Primitives;
+using Newtonsoft.Json;
+
+namespace Mvc.DynamicCorsPolicyProvider
+{
+    public class DynamicCorsPolicyProvider : ICorsPolicyProvider
+    {
+        private readonly PhysicalFileProvider _fileProvider;
+        private readonly ILogger<DynamicCorsPolicyProvider> _logger;
+        private Dictionary<string, CorsPolicy> _policies = null;
+
+        public DynamicCorsPolicyProvider(PhysicalFileProvider fileProvider, ILogger<DynamicCorsPolicyProvider> logger)
+        {
+            _fileProvider = fileProvider;
+            _logger = logger;
+
+            // Initialize the policies the first time this policy provider is instantiated (singleton)
+            BuildPolicies();
+
+            // Register a file change callback to refresh these policies whenever a change is made
+            RegisterCallback();
+        }
+
+        public Task<CorsPolicy> GetPolicyAsync(HttpContext context, string policyName)
+        {
+            return Task.FromResult(_policies[policyName]);
+        }
+
+        private void RegisterCallback()
+        {
+            _fileProvider.Watch("corsSettings.json")
+                .RegisterChangeCallback((state) =>
+                {
+                    _logger.LogInformation("CORS settings changed. Updating policies...");
+                    BuildPolicies();
+
+                    // Re-register callback
+                    RegisterCallback();
+                }, state: null);
+        }
+
+        private void BuildPolicies()
+        {
+            var content = File.ReadAllText(_fileProvider.GetFileInfo("corsSettings.json").PhysicalPath);
```

Comment:
Can call `.CreateReadStream()` on the result of `GetFileInfo` and thus use the abstraction (there might not be a `PhysicalPath`).